aws/nat_gateway.go

```diff
@@ -0,0 +1,156 @@
+package aws
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/ec2"
+	"github.com/gruntwork-io/go-commons/errors"
+	"github.com/gruntwork-io/go-commons/retry"
+	multierror "github.com/hashicorp/go-multierror"
+
+	"github.com/gruntwork-io/cloud-nuke/config"
+	"github.com/gruntwork-io/cloud-nuke/logging"
+)
+
+func getAllNatGateways(session *session.Session, excludeAfter time.Time, configObj config.Config) ([]*string, error) {
+	svc := ec2.New(session)
+
+	allNatGateways := []*string{}
+	input := &ec2.DescribeNatGatewaysInput{}
+	err := svc.DescribeNatGatewaysPages(
+		input,
+		func(page *ec2.DescribeNatGatewaysOutput, lastPage bool) bool {
+			for _, ngw := range page.NatGateways {
+				if shouldIncludeNatGateway(ngw, excludeAfter, configObj) {
+					allNatGateways = append(allNatGateways, ngw.NatGatewayId)
+				}
+			}
+			return !lastPage
+		},
+	)
+	return allNatGateways, errors.WithStackTrace(err)
+}
+
+func shouldIncludeNatGateway(ngw *ec2.NatGateway, excludeAfter time.Time, configObj config.Config) bool {
+	if ngw == nil {
+		return false
+	}
+
+	if ngw.CreateTime != nil && excludeAfter.Before(*ngw.CreateTime) {
+		return false
+	}
+
+	return config.ShouldInclude(
+		getNatGatewayName(ngw),
+		configObj.NatGateway.IncludeRule.NamesRegExp,
+		configObj.NatGateway.ExcludeRule.NamesRegExp,
+	)
+}
+
+func getNatGatewayName(ngw *ec2.NatGateway) string {
+	for _, tag := range ngw.Tags {
+		if aws.StringValue(tag.Key) == "Name" {
+			return aws.StringValue(tag.Value)
+		}
+	}
+	return ""
+}
+
+func nukeAllNatGateways(session *session.Session, identifiers []*string) error {
+	region := aws.StringValue(session.Config.Region)
+
+	svc := ec2.New(session)
+
+	if len(identifiers) == 0 {
+		logging.Logger.Infof("No Nat Gateways to nuke in region %s", region)
+		return nil
+	}
+
+	// There is no bulk delete nat gateway API, so we delete the batch of nat gateways concurrently using go routines.
+	logging.Logger.Infof("Deleting Nat Gateways in region %s", region)
+	wg := new(sync.WaitGroup)
+	wg.Add(len(identifiers))
+	errChans := make([]chan error, len(identifiers))
+	for i, ngwID := range identifiers {
+		errChans[i] = make(chan error, 1)
+		go deleteNatGatewayAsync(wg, errChans[i], svc, ngwID)
```

Comment:
NIT: Does AWS throttle us if we do too many of these concurrently?

Reply:
Added a guard to be extra clear: 039adb4788bb302f4447c87ce05f6bdb56b1d650